src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1255 @@
+        if (r < 0)
+                return 0; /* map_all_properties does the debug logging internally... */
+
+        if (progress == UINT_MAX)
+                return 0;
+
+        r = ordered_hashmap_replace(map, op->target_id, INT_TO_PTR((int) progress));
+        if (r < 0)
+                log_debug_errno(r, "Failed to update hashmap: %m");
+        return 0;
+}
+
+static int update_finished(sd_bus_message *m, void *userdata, sd_bus_error *error) {
+        _cleanup_(operation_freep) Operation *op = ASSERT_PTR(userdata);
+        OrderedHashmap *map = ASSERT_PTR(op->userdata);
+        uint64_t id;
+        int r, status;
+
+        assert(m);
+
+        r = sd_bus_message_read(m, "toi", &id, NULL, &status);
+        if (r < 0) {
+                bus_log_parse_error_debug(r);
+                return 0;
+        }
+
+        if (id != op->job_id) {
+                TAKE_PTR(op);
+                return 0;
+        }
+
+        if (status == 0) /* success */
+                status = 100;
+        else if (status > 0) /* exit status without errno */
+                status = UPDATE_PROGRESS_FAILED; /* i.e. EXIT_FAILURE */
+        /* else errno */
+
+        r = ordered_hashmap_replace(map, op->target_id, INT_TO_PTR(status));
+        if (r < 0)
+                log_debug_errno(r, "Failed to update hashmap: %m");
+        return 0;
+}
+
+static int update_interrupted(sd_event_source *source, void *userdata) {
+        /* Since the event loop is exiting, we will never recieve the JobRemoved
+         * signal. So, we must free the userdata here. */
+        _cleanup_(operation_freep) Operation *op = ASSERT_PTR(userdata);
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        OrderedHashmap *map = ASSERT_PTR(op->userdata);
+        int r;
+
+        r = sd_bus_call_method(op->bus,
+                               bus_sysupdate_mgr->destination,
+                               op->job_path,
+                               "org.freedesktop.sysupdate1.Job",
+                               "Cancel",
+                               &error, /* reply= */ NULL,
+                               NULL);
+        if (r < 0)
+                return log_bus_error(r, &error, NULL, "call Cancel");
+
+        r = ordered_hashmap_replace(map, op->target_id, INT_TO_PTR(-ECANCELED));
+        if (r < 0)
+                log_debug_errno(r, "Failed to update hashmap: %m");
+
+        return 0;
+}
+
+static int update_started(sd_bus_message *reply, void *userdata, sd_bus_error *ret_error) {
+        _cleanup_(operation_freep) Operation *op = ASSERT_PTR(userdata);
+        OrderedHashmap *map = ASSERT_PTR(op->userdata);
+        const sd_bus_error *e;
+        _cleanup_free_ char *key = NULL;
+        const char *new_version, *job_path;
+        int r;
+
+        assert(reply);
+
+        e = sd_bus_message_get_error(reply);
+        if (e) {
+                r = -sd_bus_error_get_errno(e);
```

Comment:
you are initializing `r` here, but then overriding it with something else before using it, see below